File: rust/src/bin/structual-patterns/adapter.rs
trait Target {
    fn request(&self) -> String;
}

pub struct OrdinaryTarget;

impl Target for OrdinaryTarget {
    fn request(&self) -> String {
        "Ordinary request".to_string()
    }
}

struct Adaptee;

impl Adaptee {
    fn specific_request(&self) -> String {
        ".tseuqer cificepS".to_string()
    }
}

struct Adapter {
    adaptee: Adaptee
}

impl Adapter {
    fn new(adaptee: Adaptee) -> Self {
        Self { adaptee }
    }
}

impl Target for Adapter {
    fn request(&self) -> String {
        self.adaptee.specific_request().chars().rev().collect::<String>()
    }
}

fn call<T: Target>(target: T) {
    println!("{}", target.request());
}

fn main() {
    let target = OrdinaryTarget;
    call(target);

    let adaptee = Adaptee;
    println!("{}", adaptee.specific_request());

    let adapter = Adapter::new(adaptee);
    call(adapter);
}

File: typescript/structual-patterns/adapter.ts
class Target {
  public request(): string {
    return "Target: The default behavior.";
  }
}

class Adaptee {
  public specificRequest(): string {
    return ".eetpadA eht fo tcerroc a si sihT";
  }
}

class Adapter extends Target {
  private adaptee: Adaptee;

  constructor(adaptee: Adaptee) {
    super();
    this.adaptee = adaptee;
  }

  public request(): string {
    const result = this.adaptee.specificRequest();

    return `Adapter: (TRANSLATED) ${result.split("").reverse().join("")}`;
  }
}

function clientCode(target: Target) {
  console.log(target.request());
}

const target = new Target();
clientCode(target);

const adaptee = new Adaptee();
console.log("Adaptee: " + adaptee.specificRequest());

const adapter = new Adapter(adaptee);
clientCode(adapter);
